src/pkg/storage/s3.go

```diff
@@ -100,3 +121,107 @@ func (c S3Config) validate() error {
 
 	return nil
 }
+
+func s3Overrides(in map[string]string) map[string]string {
+	return map[string]string{
+		Bucket:                 in[Bucket],
+		Endpoint:               in[Endpoint],
+		Prefix:                 in[Prefix],
+		DoNotUseTLS:            in[DoNotUseTLS],
+		DoNotVerifyTLS:         in[DoNotVerifyTLS],
+		StorageProviderTypeKey: in[StorageProviderTypeKey],
+	}
+}
+
+func (c *S3Config) s3ConfigsFromStore(kvg KVStoreGetter) {
+	c.Bucket = cast.ToString(kvg.Get(BucketNameKey))
+	c.Endpoint = cast.ToString(kvg.Get(EndpointKey))
+	c.Prefix = cast.ToString(kvg.Get(PrefixKey))
+	c.DoNotUseTLS = cast.ToBool(kvg.Get(DisableTLSKey))
+	c.DoNotVerifyTLS = cast.ToBool(kvg.Get(DisableTLSVerificationKey))
+}
+
+func (c *S3Config) s3CredsFromStore(kvg KVStoreGetter) {
+	c.AccessKey = cast.ToString(kvg.Get(AccessKey))
+	c.SecretKey = cast.ToString(kvg.Get(SecretAccessKey))
+	c.SessionToken = cast.ToString(kvg.Get(SessionToken))
+}
+
+var _ Configurer = &S3Config{}
+
+func (c *S3Config) ApplyConfigOverrides(
+	kvg KVStoreGetter,
+	readConfigFromStore bool,
+	matchFromConfig bool,
+	overrides map[string]string,
+) error {
+	if readConfigFromStore {
+		c.s3ConfigsFromStore(kvg)
+
+		if b, ok := overrides[Bucket]; ok {
+			overrides[Bucket] = common.NormalizeBucket(b)
+		}
+
+		if p, ok := overrides[Prefix]; ok {
+			overrides[Prefix] = common.NormalizePrefix(p)
+		}
```

Comment:
so long as these funcs return an empty string when given an empty string, the conditionals are unnecessary.  Golang will automatically return an empty string if the key is not populated.
```suggestion
overrides[Bucket] = common.NormalizeBucket(overrides[Bucket])
overrides[Prefix] = common.NormalizePrefix(overrides[Prefix])
```